mypy/checker.py

```diff
@@ -4865,6 +4860,56 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
             expr = parent_expr
             expr_type = output[parent_expr] = make_simplified_union(new_parent_types)
 
+    def refine_optional_in(self,
```

Comment:
The narrowing support provided by this commit is very cool, but I don't have enough brainpower this evening to look at all the new code in this commit in detail.

**Requesting a second pair of eyes on the narrowing logic in this commit.**

Reply:
Did review the narrowing support here. No further review is requested.